// @adrise/web-ui/lib/Input/Input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Icons = require("../Icons/Icons");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

/**
 * Input Field, including `input`, `textarea` and `select`
 */
var Input =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Input, _PureComponent);

  function Input(props) {
    var _this;

    _classCallCheck(this, Input);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Input).call(this, props));

    _this.handleFocus = function (e) {
      var onFocus = _this.props.onFocus;

      if (onFocus) {
        onFocus(e);
      }

      _this.setState({
        active: true
      });
    };

    _this.handleBlur = function (e) {
      var onBlur = _this.props.onBlur;

      if (onBlur) {
        onBlur(e);
      }

      _this.setState({
        active: false
      });
    };

    _this.handleChange = function (e) {
      var onChange = _this.props.onChange;

      if (onChange) {
        onChange(e);
      }

      _this.setState({
        active: true
      });
    };

    _this.togglePasswordVisible = function () {
      _this.setState({
        passwordVisible: !_this.state.passwordVisible
      });
    };

    _this.state = {
      active: false,
      passwordVisible: false
    };
    return _this;
  }

  _createClass(Input, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          Tag = _this$props.tag,
          type = _this$props.type,
          label = _this$props.label,
          hint = _this$props.hint,
          maxLength = _this$props.maxLength,
          className = _this$props.className,
          canPasswordVisible = _this$props.canPasswordVisible,
          fixedLabel = _this$props.fixedLabel,
          disabled = _this$props.disabled,
          showCounter = _this$props.showCounter,
          value = _this$props.value,
          error = _this$props.error,
          name = _this$props.name,
          others = _objectWithoutProperties(_this$props, ["tag", "type", "label", "hint", "maxLength", "className", "canPasswordVisible", "fixedLabel", "disabled", "showCounter", "value", "error", "name"]);

      var _this$state = this.state,
          active = _this$state.active,
          passwordVisible = _this$state.passwordVisible;
      var finalType = type === 'password' && canPasswordVisible && passwordVisible ? 'text' : type;
      var classes = (0, _classnames.default)('Input', {
        'Input--disabled': disabled,
        'Input--actived': active,
        'Input--fixed': fixedLabel,
        'Input--errored': error,
        'Input--filled': value,
        'Input--textarea': Tag === 'textarea'
      }, className);
      var passwordIcon = null;

      if (type === 'password' && canPasswordVisible && value) {
        passwordIcon = passwordVisible ? _react.default.createElement(_Icons.Hide, {
          className: "Input__password-icon",
          onClick: this.togglePasswordVisible
        }) : _react.default.createElement(_Icons.Reveal, {
          className: "Input__password-icon",
          onClick: this.togglePasswordVisible
        });
      }

      return _react.default.createElement("div", {
        className: classes
      }, _react.default.createElement(Tag, _extends({
        className: "Input__input",
        name: name,
        type: finalType,
        value: value,
        maxLength: maxLength,
        disabled: disabled
      }, others, {
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        onChange: this.handleChange
      })), error || label ? _react.default.createElement("span", {
        className: "Input__label"
      }, error || label) : null, maxLength && showCounter && value ? _react.default.createElement("span", {
        className: "Input__counter"
      }, (value || '').length, "/", maxLength) : null, hint ? _react.default.createElement("span", {
        className: "Input__hint"
      }, hint) : null, passwordIcon);
    }
  }]);

  return Input;
}(_react.PureComponent);

exports.default = Input;
Input.propTypes = {
  /** input tag name */
  tag: _propTypes.default.oneOf(['input', 'textarea', 'select']),

  /** input type */
  type: _propTypes.default.string,

  /** name for identifying with Formik.js */
  name: _propTypes.default.string,

  /** label text */
  label: _propTypes.default.string,

  /** hint text */
  hint: _propTypes.default.string,

  /** max length of input text */
  maxLength: _propTypes.default.number,

  /** whether password can be visible */
  canPasswordVisible: _propTypes.default.bool,

  /** whether label is fixed or floating */
  fixedLabel: _propTypes.default.bool,

  /** whether letter counter is displayed */
  showCounter: _propTypes.default.bool,

  /** whether this field is disabled */
  disabled: _propTypes.default.bool,

  /** callback when this field receives focus */
  onFocus: _propTypes.default.func,

  /** callback when this field loses focus */
  onBlur: _propTypes.default.func,

  /** callback when this field's value changes */
  onChange: _propTypes.default.func,

  /** for overriding the original input style */
  className: _propTypes.default.string,

  /** input value */
  value: _propTypes.default.string,

  /** error message */
  error: _propTypes.default.string
};
Input.defaultProps = {
  tag: 'input',
  type: 'text',
  canPasswordVisible: true,
  showCounter: true
};
