// @adrise/web-ui/lib/Carousel/Carousel.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _resizeObserverPolyfill = _interopRequireDefault(require("resize-observer-polyfill"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _Icons = require("../Icons/Icons");

var _Row = _interopRequireDefault(require("../Row/Row"));

var _debounce = require("../utils/debounce");

var _tools = require("../utils/tools");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var count = _react.Children.count;

var noOp = function noOp() {};
/**
 * Carousel is responsible for managing the layout/UX of any list of tiles.
 * Cooperated with `Col`, it's quite easy to setup responsible slideshows.
 */


var Carousel =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Carousel, _PureComponent);

  function Carousel(props) {
    var _this;

    _classCallCheck(this, Carousel);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Carousel).call(this, props));

    _this.onTouchStart = function (event) {
      if (!event.touches) return;
      var _event$touches$ = event.touches[0],
          pageX = _event$touches$.pageX,
          pageY = _event$touches$.pageY;
      _this.touchInfo = {
        startX: pageX,
        startY: pageY
      };
    };

    _this.onTouchMove = function (event) {
      var _event$touches$2 = event.touches[0],
          endX = _event$touches$2.pageX,
          endY = _event$touches$2.pageY;
      var _this$touchInfo = _this.touchInfo,
          startX = _this$touchInfo.startX,
          startY = _this$touchInfo.startY;
      var direction = (0, _tools.getSwipeDirection)(startX, startY, endX, endY); // prevent default if is an valid swipe

      if (direction !== 0) {
        event.preventDefault();
      }

      _this.touchInfo = _objectSpread({}, _this.touchInfo, {
        endX: endX,
        endY: endY,
        direction: direction
      });
    };

    _this.onTouchEnd = function () {
      _this.handleSwipe();
    };

    _this.onTouchCancel = function () {
      _this.touchInfo = {};
    };

    _this.previous = function () {
      var _this$state = _this.state,
          index = _this$state.index,
          colsPerPage = _this$state.colsPerPage;
      var newIndex = Math.max(index - colsPerPage, 0);
      var newPageIndex = Math.floor((newIndex + colsPerPage - 1) / colsPerPage);

      _this.trackTriggers({
        pageIndex: newPageIndex,
        itemIndex: newIndex
      });

      _this.setState({
        index: newIndex
      });
    };

    _this.next = function () {
      var _this$props = _this.props,
          loadMoreItems = _this$props.loadMoreItems,
          advance = _this$props.advance,
          children = _this$props.children;
      var _this$state2 = _this.state,
          index = _this$state2.index,
          colsPerPage = _this$state2.colsPerPage;
      var newIndex = advance ? Math.max(0, Math.min(index + colsPerPage, count(children) - colsPerPage)) : index + colsPerPage;

      if (loadMoreItems) {
        loadMoreItems(colsPerPage, newIndex);
      } // eslint-disable-next-line no-mixed-operators


      var newPageIndex = Math.floor((newIndex + colsPerPage - 1) / colsPerPage);

      _this.setState({
        index: newIndex
      });

      _this.trackTriggers({
        pageIndex: newPageIndex,
        itemIndex: newIndex
      });
    };

    _this.state = {
      index: props.initialIndex || 0,
      colsPerPage: 0,
      colWidth: 0
    }; // we set transition to true in CDU

    _this.transition = false;
    _this.touchInfo = {};
    _this.resizeTimer = null;
    _this.debouncedResizeHandler = (0, _debounce.debounce)(function () {
      return _this.resizeHandler();
    }, 250); // set ref to container div

    _this.carouselContainer = _react.default.createRef(); // add resizeObserver class prop

    _this.resizeObserver = null;
    _this.resizeObserver = new _resizeObserverPolyfill.default(_this.debouncedResizeHandler);
    return _this;
  }

  _createClass(Carousel, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (this.resizeObserver && this.carouselContainer) {
        this.resizeObserver.observe(this.carouselContainer.current);
      }

      this.debouncedResizeHandler();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (count(prevProps.children) !== count(this.props.children)) {
        this.debouncedResizeHandler();
      }

      if (!this.transition && this.state.colWidth > 0) {
        this.transition = true;
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (this.carouselContainer) {
        this.resizeObserver.unobserve(this.carouselContainer.current);
      }

      this.debouncedResizeHandler.clear();
    }
  }, {
    key: "handleSwipe",
    value: function handleSwipe() {
      var direction = this.touchInfo.direction;

      if (direction === 1) {
        this.next();
      } else if (direction === -1) {
        this.previous();
      }

      this.touchInfo = {};
    }
  }, {
    key: "resizeHandler",
    value: function resizeHandler() {
      var _this2 = this;

      var colCount = count(this.props.children);
      var colWidth = this.el.scrollWidth / colCount;
      this.setState(function (s) {
        return _objectSpread({}, s, {
          colsPerPage: Math.round(_this2.el.getBoundingClientRect().width / colWidth),
          colWidth: colWidth
        });
      });
    }
  }, {
    key: "trackTriggers",

    /**
     * track trigger click
     * @param {Object} pageIndex the destination page index, 0-based,
     * itemIndex is the index within 'children' on the left edge, 0-based
     */
    value: function trackTriggers(_ref) {
      var pageIndex = _ref.pageIndex,
          itemIndex = _ref.itemIndex;
      var onTriggerClick = this.props.onTriggerClick;

      if (onTriggerClick) {
        // the tracking system use 1-based index
        onTriggerClick({
          pageIndex: pageIndex + 1,
          itemIndex: itemIndex
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      /* eslint-disable jsx-a11y/no-static-element-interactions */
      var _this$props2 = this.props,
          children = _this$props2.children,
          noMask = _this$props2.noMask,
          className = _this$props2.className,
          theme = _this$props2.theme,
          initialIndex = _this$props2.initialIndex,
          others = _objectWithoutProperties(_this$props2, ["children", "noMask", "className", "theme", "initialIndex"]);

      var _this$state3 = this.state,
          index = _this$state3.index,
          colWidth = _this$state3.colWidth,
          colsPerPage = _this$state3.colsPerPage;
      var colCount = count(children);
      var hasPrevious = index > 0;
      var hasNext = index + colsPerPage < colCount;
      var classes = (0, _classnames.default)('Carousel', {
        'Carousel--no-mask': noMask
      }, className);
      var preClasses = (0, _classnames.default)('Carousel__prev', {
        'Carousel__arrow-active': hasPrevious
      });
      var preMaskClasses = (0, _classnames.default)('Carousel__prev-mask', {
        'Carousel_light-theme': theme === 'light'
      });
      var nextClasses = (0, _classnames.default)('Carousel__next', {
        'Carousel__arrow-active': hasNext
      });
      var nextMaskClasses = (0, _classnames.default)('Carousel__next-mask', {
        'Carousel_light-theme': theme === 'light'
      });
      var rowStyle = {
        transform: "translate3d(".concat(-index * colWidth, "px, 0, 0)")
      };
      if (!this.transition) rowStyle.transition = 'none';
      return _react.default.createElement("div", _extends({
        ref: this.carouselContainer,
        className: classes,
        onTouchStart: this.onTouchStart,
        onTouchMove: this.onTouchMove,
        onTouchEnd: this.onTouchEnd,
        onTouchCancel: this.onTouchCancel
      }, (0, _tools.omit)(others, ['onTriggerClick', 'loadMoreItems', 'advance'])), _react.default.createElement(_Button.default, {
        shape: "round",
        color: "secondary",
        className: preClasses,
        onClick: this.previous
      }, _react.default.createElement(_Icons.CaretArrow, null)), _react.default.createElement(_Button.default, {
        shape: "round",
        color: "secondary",
        className: nextClasses,
        onClick: this.next
      }, _react.default.createElement(_Icons.CaretArrow, null)), noMask || !hasPrevious ? null : _react.default.createElement("div", {
        className: preMaskClasses,
        onClick: this.previous,
        onKeyPress: noOp
      }), noMask || !hasNext ? null : _react.default.createElement("div", {
        className: nextMaskClasses,
        onClick: this.next,
        onKeyPress: noOp
      }), _react.default.createElement("div", {
        className: "Carousel__content"
      }, _react.default.createElement(_Row.default, {
        nodeRef: function nodeRef(el) {
          _this3.el = el;
        },
        className: "Carousel__row",
        style: rowStyle
      }, _react.default.Children.toArray(children))));
    }
  }]);

  return Carousel;
}(_react.PureComponent);

exports.default = Carousel;
Carousel.propTypes = {
  /** load more items callback when clicking next */
  loadMoreItems: _propTypes.default.func,

  /** trigger tracking callback when clicking next, receive a 1-based page index parameter */
  onTriggerClick: _propTypes.default.func,

  /** not show mask */
  noMask: _propTypes.default.bool,

  /** fill the last page full */
  advance: _propTypes.default.bool,

  /** for overriding the original input style */
  className: _propTypes.default.string,

  /** tiles, each tile should be wrapped by `Col` */
  children: _propTypes.default.node.isRequired,

  /** useful for situations such as initial in view of current episode in episodes Carousel */
  initialIndex: _propTypes.default.number,

  /** steel-grey or light */
  theme: _propTypes.default.oneOf(['steel-grey', 'light'])
};
Carousel.defaultProps = {
  noMask: false,
  advance: false,
  initialIndex: 0,
  theme: 'steel-grey'
};
