// @adrise/web-ui/lib/Select/Select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _tools = require("../utils/tools");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var noOp = function noOp() {};

var KeyCode = {
  ENTER: 13,
  UP: 38,
  DOWN: 40
};
/**
 * A custom select component to meet our brand design
 */

var Select =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(Select, _PureComponent);

  _createClass(Select, null, [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      if (props.disabled && state.active) {
        return {
          active: false
        };
      }

      return null;
    }
  }]);

  function Select(props) {
    var _this;

    _classCallCheck(this, Select);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Select).call(this, props));

    _this.handleClickWhenDropDownOpen = function (domEvent) {
      var target = domEvent.target;
      var selectNode = _this.node; // the clicked node is a child of selectNode
      // @link https://www.w3schools.com/jsref/met_node_comparedocumentposition.asp
      // using compareDocumentPosition since it has a better support than contains

      if (selectNode.compareDocumentPosition(target) === 20) {
        var _ref = target || {},
            _ref$dataset = _ref.dataset,
            dataset = _ref$dataset === void 0 ? {} : _ref$dataset;

        var value = dataset.value; // it is an list item
        // we are ok with getting empty string or 0 as value

        if (typeof value !== 'undefined') {
          // create an event type object with target.value. also pass along the original event object
          // as second argument in case it would be needed
          _this.handleChange({
            target: {
              value: value
            }
          }, domEvent);
        }
      }

      _this.removeClickEventListener(); // not all browsers support once: true


      _this.setState({
        active: false
      });
    };

    _this.handleKeyDownWhenDropDownOpen = function (e) {
      e.preventDefault();
      var key = e.which || e.keyCode;
      var options = _this.props.options;
      var highlightIndex = _this.state.highlightIndex;

      switch (key) {
        case KeyCode.UP:
          {
            var prevIndex = highlightIndex - 1;

            if (options[prevIndex]) {
              _this.scrollToOption(prevIndex);

              _this.setState({
                highlightIndex: prevIndex
              });
            }

            break;
          }

        case KeyCode.DOWN:
          {
            var nextIndex = highlightIndex + 1;

            if (options[nextIndex]) {
              _this.scrollToOption(nextIndex);

              _this.setState({
                highlightIndex: nextIndex
              });
            }

            break;
          }

        case KeyCode.ENTER:
          {
            if (highlightIndex >= 0 && highlightIndex < options.length) {
              var value = options[highlightIndex].value;

              _this.handleChange({
                target: {
                  value: value
                }
              }, e);

              _this.setState({
                active: false
              });
            }

            break;
          }

        default:
          {
            break;
          }
      }
    };

    _this.scrollToOption = function (index) {
      var wrapper = _this.node.getElementsByTagName('ul')[0];

      var optionNodes = _this.node.getElementsByTagName('li');

      var candidate = optionNodes[index];

      var _candidate$getBoundin = candidate.getBoundingClientRect(),
          height = _candidate$getBoundin.height;

      if (candidate.offsetTop < wrapper.scrollTop) {
        wrapper.scrollTop = candidate.offsetTop;
      } else if (candidate.offsetTop + height > wrapper.scrollTop + wrapper.clientHeight) {
        wrapper.scrollTop = candidate.offsetTop + height - wrapper.clientHeight;
      }
    };

    _this.removeKeyDownEvent = function () {
      window.removeEventListener('keydown', _this.handleKeyDownWhenDropDownOpen);
    };

    _this.handleClick = function () {
      if (_this.props.disabled) return;

      _this.setState({
        active: true
      });
    };

    _this.handleFocus = function (e) {
      var _this$props = _this.props,
          onFocus = _this$props.onFocus,
          disabled = _this$props.disabled;
      if (disabled) return;
      /* istanbul ignore else */

      if (onFocus) onFocus(e);

      _this.setState({
        active: true
      });
    };

    _this.handleBlur = function (e) {
      var _this$props2 = _this.props,
          onBlur = _this$props2.onBlur,
          disabled = _this$props2.disabled;
      if (disabled) return;
      /* istanbul ignore else */

      if (onBlur) onBlur(e);

      _this.setState({
        active: false
      });
    };

    _this.handleChange = function (e, domEvent) {
      var onChange = _this.props.onChange;
      /* istanbul ignore else */

      if (onChange) onChange(e, domEvent);
    };

    _this.nativeOnChange = function (e) {
      _this.setState({
        active: false
      });

      _this.props.onChange(e);
    };

    _this.renderDesktopSelect = function () {
      var _this$state = _this.state,
          active = _this$state.active,
          highlightIndex = _this$state.highlightIndex;
      var _this$props3 = _this.props,
          options = _this$props3.options,
          selectValue = _this$props3.value;
      if (!active) return null;
      return _react.default.createElement("ul", {
        className: "Select__list",
        role: "listbox"
      }, options.map(function (option, i) {
        var value = option.value,
            label = option.label,
            optionStyle = option.optionStyle;
        var isSelected = selectValue === value;
        var listItemClasses = (0, _classnames.default)('Select__listItem', {
          'Select__listItem--selected': isSelected,
          'Select__listItem--highlighted': i === highlightIndex
        });
        return _react.default.createElement("li", {
          "data-value": value,
          key: value,
          className: listItemClasses,
          style: optionStyle,
          role: "option",
          "aria-selected": isSelected
        }, label || value);
      }));
    };

    _this.renderMobileSelect = function () {
      var _this$props4 = _this.props,
          options = _this$props4.options,
          selectValue = _this$props4.value,
          name = _this$props4.name;
      return _react.default.createElement("select", {
        name: name,
        value: selectValue,
        className: "Select__native",
        onChange: _this.nativeOnChange,
        tabIndex: 0
      }, options.map(function (option) {
        var value = option.value,
            label = option.label;
        return _react.default.createElement("option", {
          key: value,
          value: value
        }, label || value);
      }));
    };

    _this.state = {
      active: props.autoFocus && !props.disabled,
      highlightIndex: props.options.findIndex(function (o) {
        return o.value === props.value;
      })
    };
    return _this;
  }

  _createClass(Select, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      var _this2 = this;

      if (this.props.native) return;
      var prevActiveState = prevState.active;
      var currentActiveState = this.state.active; // since div does not support focus and blur we will do this manually

      if (prevActiveState && !currentActiveState) {
        this.removeClickEventListener();
        this.removeKeyDownEvent();
      }

      if (!prevActiveState && currentActiveState) {
        var highlightIndex = this.props.options.findIndex(function (o) {
          return o.value === _this2.props.value;
        });
        /* eslint-disable-next-line react/no-did-update-set-state */

        this.setState({
          highlightIndex: highlightIndex
        });
        this.scrollToOption(highlightIndex);
        window.addEventListener('click', this.handleClickWhenDropDownOpen, {
          once: true,
          capture: true
        });
        window.addEventListener('keydown', this.handleKeyDownWhenDropDownOpen);
      }
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      /* istanbul ignore next */
      if (!this.props.native) {
        this.removeClickEventListener();
        this.removeKeyDownEvent();
      }
    }
    /**
     * finds the selected option
     */

  }, {
    key: "getSelectedOption",
    value: function getSelectedOption() {
      var _this$props5 = this.props,
          options = _this$props5.options,
          value = _this$props5.value;
      var index = (0, _tools.findIndex)(options, function (option) {
        return option.value === value;
      });
      return options[index];
    }
    /* istanbul ignore next */

  }, {
    key: "removeClickEventListener",
    value: function removeClickEventListener() {
      /* istanbul ignore next */
      window.removeEventListener('click', this.handleClickWhenDropDownOpen, {
        once: true,
        capture: true
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props6 = this.props,
          disabled = _this$props6.disabled,
          fixedLabel = _this$props6.fixedLabel,
          error = _this$props6.error,
          className = _this$props6.className,
          selectValue = _this$props6.value,
          selectLabel = _this$props6.label,
          name = _this$props6.name,
          native = _this$props6.native;
      var active = this.state.active;
      var erroring = error && !active;
      var classes = (0, _classnames.default)('Input', 'Select', {
        'Input--disabled': disabled,
        'Input--fixed': fixedLabel,
        'Input--errored': erroring,
        'Input--filled': selectValue
      }, className);
      var downIconClasses = (0, _classnames.default)('Select__down-icon', {
        'Select__down-icon--active': active
      });
      var selectedOption = this.getSelectedOption();
      return _react.default.createElement("div", {
        id: name,
        tabIndex: native ? -1 : 0,
        className: classes,
        onBlur: this.handleBlur,
        onFocus: this.handleFocus,
        onClick: this.handleClick,
        onKeyPress: noOp,
        "aria-haspopup": "listbox",
        role: "button",
        ref: function ref(node) {
          _this3.node = node;
        }
      }, selectedOption && typeof selectedOption.label !== 'undefined' ? _react.default.createElement("span", {
        className: "Input__input",
        style: selectedOption.optionStyle
      }, selectedOption.label) : null, erroring ? _react.default.createElement("span", {
        className: "Input__label"
      }, error) : null, !erroring && selectLabel ? _react.default.createElement("span", {
        className: "Input__label"
      }, selectLabel) : null, _react.default.createElement("span", {
        className: downIconClasses
      }), native ? this.renderMobileSelect() : this.renderDesktopSelect());
    }
  }]);

  return Select;
}(_react.PureComponent);

exports.default = Select;
Select.propTypes = {
  name: _propTypes.default.string,
  value: _propTypes.default.string,
  // value of the currently selected option
  onChange: _propTypes.default.func.isRequired,
  options: _propTypes.default.arrayOf(_propTypes.default.shape({
    value: _propTypes.default.string.isRequired,
    label: _propTypes.default.string,
    optionStyle: _propTypes.default.object
  })).isRequired,
  onBlur: _propTypes.default.func,
  label: _propTypes.default.string,
  // label of the select e.g. Gender
  error: _propTypes.default.string,
  onFocus: _propTypes.default.func,
  disabled: _propTypes.default.bool,
  autoFocus: _propTypes.default.bool,
  fixedLabel: _propTypes.default.bool,
  className: _propTypes.default.string,
  native: _propTypes.default.bool
};
Select.defaultProps = {
  autoFocus: false,
  disabled: false,
  name: 'select-input',
  native: true // defaulting to native as it is safer

};
