// @adrise/web-ui/lib/Button/Button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var BUTTON_COLORS = {
  primary: 'primary',
  secondary: 'secondary',
  transparent: 'transparent'
};
var BASENAME = 'Button';
/**
 * "An elegant button, for a more civilized age."
 * Trying to be simpler than our current button
 * with some strict styling choices.
 */

var Button = function Button(_ref) {
  var _classNames;

  var color = _ref.color,
      size = _ref.size,
      shape = _ref.shape,
      inverse = _ref.inverse,
      block = _ref.block,
      icon = _ref.icon,
      children = _ref.children,
      className = _ref.className,
      onClick = _ref.onClick,
      others = _objectWithoutProperties(_ref, ["color", "size", "shape", "inverse", "block", "icon", "children", "className", "onClick"]);

  var _Button$defaultProps = Button.defaultProps,
      defaultSize = _Button$defaultProps.size,
      defaultColor = _Button$defaultProps.color;
  var classes = (0, _classnames.default)(BASENAME, (_classNames = {}, _defineProperty(_classNames, "".concat(BASENAME, "--").concat(size), size !== defaultSize), _defineProperty(_classNames, "".concat(BASENAME, "--").concat(color), color && BUTTON_COLORS[color] && color !== defaultColor), _defineProperty(_classNames, "".concat(BASENAME, "--").concat(shape), shape), _defineProperty(_classNames, "".concat(BASENAME, "--only-icon"), icon && !children), _defineProperty(_classNames, "".concat(BASENAME, "--block"), block), _defineProperty(_classNames, "".concat(BASENAME, "--inverse"), inverse), _classNames), className);
  return _react.default.createElement("button", _extends({
    className: classes,
    onClick: onClick
  }, others), _react.default.createElement("div", {
    className: "".concat(BASENAME, "__bg"),
    style: color && !BUTTON_COLORS[color] ? {
      background: color
    } : null
  }), _react.default.createElement("div", {
    className: "".concat(BASENAME, "__content")
  }, icon ? _react.default.createElement("i", {
    className: "".concat(BASENAME, "__icon")
  }, icon) : null, children));
};

Button.propTypes = {
  /** background color. 3 builtin choices, also support customized color */
  color: _propTypes.default.string,

  /** main size */
  size: _propTypes.default.oneOf(['small', 'normal', 'large']),

  /** button outline shape */
  shape: _propTypes.default.oneOf(['round']),

  /** light background */
  inverse: _propTypes.default.bool,

  /** give width 100% */
  block: _propTypes.default.bool,

  /** if you want to add an `<SVG/>` */
  icon: _propTypes.default.element,

  /** custom class for wrapper */
  className: _propTypes.default.string,

  /** function to call on click action */
  onClick: _propTypes.default.func,

  /** what is inside your button */
  children: _propTypes.default.node
};
Button.defaultProps = {
  color: BUTTON_COLORS.primary,
  size: 'normal',
  inverse: false,
  block: false
};
var _default = Button;
exports.default = _default;
