// @adrise/web-ui/lib/utils/tools.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findIndex = findIndex;
exports.getSwipeDirection = exports.getSwipeDistance = exports.omit = exports.isObject = exports.isUndefined = void 0;

var _constants = require("../constants/constants");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/**
 * whether the param is `undefined`
 * @param {*} obj
 * @return {Boolean}
 */
var isUndefined = function isUndefined(obj) {
  return typeof obj === 'undefined';
};
/**
 * whether the param is an object
 * @param {*} obj
 * @return {Boolean}
 */


exports.isUndefined = isUndefined;

var isObject = function isObject(obj) {
  return _typeof(obj) === 'object';
};
/**
 * omit specified keys from original object
 * @param {Object} obj
 * @param {[String]} keys
 * @return {Object}
 */


exports.isObject = isObject;

var omit = function omit(obj, keys) {
  return Object.keys(obj).reduce(function (acc, key) {
    if (keys.indexOf(key) === -1) {
      acc[key] = obj[key];
    }

    return acc;
  }, {});
};
/**
 * get distance of swipe
 * @param {Number} startX
 * @param {Number} startY
 * @param {Number} endX
 * @param {Number} endY
 * @returns {Number}
 */


exports.omit = omit;

var getSwipeDistance = function getSwipeDistance(startX, startY, endX, endY) {
  var xCoordSquared = Math.pow(startX - endX, 2);
  var yCoordSquared = Math.pow(startY - endY, 2);
  return Math.round(Math.sqrt(xCoordSquared + yCoordSquared));
};
/**
 * get swipe direction
 * @param {Number} startX
 * @param {Number} startY
 * @param {Number} endX
 * @param {Number} endY
 * @returns {Number} 1 for next, -1 for previous, 0 for none swipe
 * if distance swiped is less than a threshold we decide, do not count it as a swipe
 */


exports.getSwipeDistance = getSwipeDistance;

var getSwipeDirection = function getSwipeDirection(startX, startY, endX, endY) {
  var distance = getSwipeDistance(startX, startY, endX, endY);
  if (distance < _constants.SWIPE_DISTANCE_THRESHOLD) return 0; // atan2 https://en.wikipedia.org/wiki/Atan2

  var radian = Math.atan2(endY - startY, endX - startX);
  var swipeAngle = Math.round(radian * 180 / Math.PI) + (radian < 0 ? 360 : 0);
  var swipeDirection = 0;

  if (swipeAngle <= 45 || swipeAngle >= 315) {
    swipeDirection = -1;
  } else if (swipeAngle >= 135 && swipeAngle <= 225) {
    swipeDirection = 1;
  }

  return swipeDirection;
};
/**
 * find the index of element which satisfy `predicate`
 * @param arr
 * @param {function} predicate, will be passed the item
 * @returns {number}
 */


exports.getSwipeDirection = getSwipeDirection;

function findIndex(arr, predicate) {
  var idx = -1;
  arr.some(function (item, index) {
    if (predicate(item)) {
      idx = index;
      return true;
    }

    return false;
  });
  return idx;
}
